Updates to CESM-FV Reader

noribeth-m opened the pull request:
I made some updates to the CESM-FV reader to have the capability to calculate the layer thickness (dz_m) if the interface pressures (hyai, hybi) are provided in the model output. The script will now calculate the variable, 'pres_pa_int', and use this variable to calculate the interface layer height, which is then used to calculate the layer thickness. These calculations are based on the default vertical structure of CESM, where the pressure is in increasing order and altitude in decreasing order. Once this is calculated, the script reorders the vertical levels so that the surface is first. This is done by default in the original code, but I added a line so that 'pres_pa_int' is also treated this way, as the original code does this based on vertical dimension 'z', which is not what is used for 'pres_pa_int' since it has an extra vertical level (33 instead of 32). 'pres_pa_int' has a vertical dimension called 'ilev'. This is currently only used in the CESM_FV reader, but is saved in the processed dataset. Moving forward, if needed, renaming the 'ilev' dimension to something like 'z_i' may be more consistent with the monetio/melodies-monet formatting, but not necessary for now. 

In addition to this, the pyresample dependency was removed and now deals with the wrap longitudes using numpy.

Commit:
copy and modify the _calc_pressure functions

Commit:
corrected new function name _calc_pressure_i

bgaubertNCAR committed:
create pres_pa_int variable

noribeth-m committed:
Add comment about pressure int

noribeth-m committed:
Added a lazy fix (for now) to be able to read the 'hyai' and 'hybi' variables because open_mfdataset seems to be adding the 'time' dimension to them. Added 'dset_load = dset_load.rename({...})' because variables were not being saved to dset_load.

noribeth-m committed:
added interface pressure, bugs still present

noribeth-m committed:
Added capability to calculate 'dz_m' if interface pressures are available.

noribeth-m committed:
removed some unnecessary comments

zmoon commented on monetio/models/_cesm_fv_mm.py:
```diff
@@ -265,3 +326,86 @@ def _calc_hydrostatic_height(dset):
+        coords={"time": dset.time, "lev": dset.lev, "lat": dset.lat, "lon": dset.lon},
+        attrs={"description": "Layer Thickness (based on interface pressure)", "units": "m"},
+    )
+    return dz_m 
```
EOF newline
```suggestion
    return dz_m

```

zmoon commented on monetio/models/_cesm_fv_mm.py:
```diff
@@ -265,3 +326,86 @@ def _calc_hydrostatic_height(dset):
+
+    Parameters
+    ----------
+    dset: xr.Dataset
```
```suggestion
    dset : xr.Dataset
```

zmoon commented on monetio/models/_cesm_fv_mm.py:
```diff
@@ -265,3 +326,86 @@ def _calc_hydrostatic_height(dset):
+
+    Parameters
+    ----------
+    dset: xr.Dataset
```
```suggestion
    dset : xr.Dataset
```

zmoon commented on monetio/models/_cesm_fv_mm.py:
```diff
@@ -216,6 +232,52 @@ def _calc_pressure(dset):
+        raise KeyError(
+            "The model does not have the variables to calculate"
+            + "the pressure required for satellite comparison"
+            + "If the vertical coordinate is not needed, set surface_only=True"
```
```suggestion
            "The model does not have the variables to calculate "
            "the pressure required for satellite comparison. "
            "If the vertical coordinate is not needed, set surface_only=True"
```

zmoon commented on monetio/models/_cesm_fv_mm.py:
```diff
@@ -265,3 +326,86 @@ def _calc_hydrostatic_height(dset):
+    xr.DataArray
+    """
+    R = 8.314  # Pa * m3 / mol K
+    M_AIR = 0.028  # kg / mol
```
```suggestion
    M_AIR = 0.029  # kg / mol
```
at this precision, should be 0.029 for dry air, no?

zmoon commented on monetio/models/_cesm_fv_mm.py:
```diff
@@ -70,25 +69,35 @@ def open_mfdataset(
+        # calc layer thickness if hyai and hybi exist
+        if "hyai" and "hybi" in dset_load.keys():
+            dset_load["pres_pa_int"] = _calc_pressure_i(dset_load)
+            var_list.append("pres_pa_int")
```
If you only need this for calculating dz here in the reader (not for something in MELODIES MONET), then you don't have to include it in the var list and then we don't have to worry about the extra vertical dim.

noribeth-m replied:
I agree here. I removed 'pres_pa_int' from the var_list and lines 131-133, where I had reordered 'ilev'.

zmoon commented on monetio/models/_cesm_fv_mm.py:
```diff
@@ -216,6 +232,52 @@ def _calc_pressure(dset):
+
+    Parameters
+    ----------
+    dset: xr.Dataset
```
```suggestion
    dset : xr.Dataset
```

zmoon reviewed (commented):
Hi @noribeth-m, some initial comments here. Also maybe we can come up with a new test file/case by selecting a few levels from data [here](https://csl.noaa.gov/groups/csl4/modeldata/melodies-monet/data/example_model_data/cesmfv_example/) since currently we're only testing a surf-only case.

noribeth-m committed:
Update monetio/models/_cesm_fv_mm.py

Co-authored-by: Zachary Moon <zmoon92@gmail.com>

noribeth-m committed:
Update monetio/models/_cesm_fv_mm.py

Co-authored-by: Zachary Moon <zmoon92@gmail.com>

noribeth-m committed:
Update monetio/models/_cesm_fv_mm.py

Co-authored-by: Zachary Moon <zmoon92@gmail.com>

noribeth-m committed:
Update monetio/models/_cesm_fv_mm.py

Co-authored-by: Zachary Moon <zmoon92@gmail.com>

noribeth-m committed:
Update monetio/models/_cesm_fv_mm.py

Co-authored-by: Zachary Moon <zmoon92@gmail.com>

noribeth-m committed:
Update monetio/models/_cesm_fv_mm.py

Co-authored-by: Zachary Moon <zmoon92@gmail.com>

noribeth-m committed:
Removed 'pres_pa_int' from var_list

zmoon commented:
@noribeth-m the automatic formatters want some changes. If you want to address this yourself, install `pre-commit` in your Python env, then in the repo run
```
pre-commit install
pre-commit run --all-files
```
and commit the result.

Or I can do it for you if you'd prefer.

noribeth-m commented:
Hi @zmoon! I installed and ran the pre-commit. It says that the _cesm_fv_mm.py file was reformatted, but I still get "Failed" in 2 places (see image attached). I'm not sure what I'm supposed to do from here. 

![Screenshot 2025-02-25 at 1 17 31 PM](https://github.com/user-attachments/assets/7abf7b7d-175a-440d-862d-514ee910688e)

zmoon commented:
@noribeth-m if you run `pre-commit run --all-files` again do you get all green? If so you can commit the result with `git` and then the format check here in GitHub will be happy.

noribeth-m committed:
ran and passed pre-commit

noribeth-m commented:
@zmoon Thanks! That worked and the checks have passed.

zmoon commented on monetio/models/_cesm_fv_mm.py:
```diff
@@ -265,3 +321,88 @@ def _calc_hydrostatic_height(dset):
+        raise Exception(
+            "Expected default CESM behaviour:"
+            + "pressure levels should be in increasing order, height in decreasing order"
+        )
```
```suggestion
        raise ValueError(
            "Expected default CESM behaviour "
            "(pressure levels should be in increasing order, height in decreasing order)"
        )
```

zmoon commented on monetio/models/_cesm_fv_mm.py:
```diff
@@ -70,12 +68,19 @@ def open_mfdataset(
         if "Z3" not in dset_load.keys():
             warnings.warn("Geopotential height Z3 is not in model keys. Assuming hydrostatic runs")
             dset_load["Z3"] = _calc_hydrostatic_height(dset_load)
+        var_list = var_list + ["alt_msl_m_mid"]
```
It looks like `pres_pa_mid` and `alt_msl_m_mid` are also added to the var list later. So perhaps remove the two lines here. xarray seems to just ignore duplicates but it's confusing to have it added twice.

zmoon commented on monetio/models/_cesm_fv_mm.py:
```diff
@@ -70,12 +68,19 @@ def open_mfdataset(
+        var_list = var_list + ["alt_msl_m_mid"]
+
+        # calc layer thickness if hyai and hybi exist
+        if "hyai" and "hybi" in dset_load.keys():
```
As it is, it's only actually checking for `'hybi'` (since `'hyai'` is always true). This is the concise way to check for both:
```suggestion
        if {"hyai", "hybi"} <= dset_load.keys():
```

zmoon commented on monetio/models/_cesm_fv_mm.py:
```diff
@@ -70,12 +68,19 @@ def open_mfdataset(
+        if "hyai" and "hybi" in dset_load.keys():
+            dset_load["pres_pa_int"] = _calc_pressure_i(dset_load)
+            dset_load["dz_m"] = _calc_layer_thickness_i(dset_load)
+            var_list.append("dz_m")
```
You could add an `else` branch here with a warning message about skipping this calculation.

zmoon commented on monetio/models/_cesm_fv_mm.py:
```diff
@@ -265,3 +321,88 @@ def _calc_hydrostatic_height(dset):
+    and altitude in decreasing order. The code flips all the variables
+    along the 'z' dimensions at the end. 'pres_pa_int' does not
+    because it has a dimension of 'ilev' instead of 'z'. Add a correction
+    for this last part.
```
👀

zmoon reviewed (commented):
A few more comments

noribeth-m committed:
Update monetio/models/_cesm_fv_mm.py

Co-authored-by: Zachary Moon <zmoon92@gmail.com>

noribeth-m committed:
Update monetio/models/_cesm_fv_mm.py

Co-authored-by: Zachary Moon <zmoon92@gmail.com>

noribeth-m committed:
Made modifications based on pull request.
    - Removed the duplicate addition of pres_pa_mid and alt_msl_m_mid to var_list.
    - Added an else statement on skipping dz_m calc if hyai and hybi not present.
    - Removed unneccessary comments.

zmoon commented on monetio/models/_cesm_fv_mm.py:
```diff
@@ -183,9 +195,9 @@ def _calc_pressure(dset):
-            + "the pressure. This can be done either with PMID or with"
-            + "P0, PS, hyam and hybm.\n"
-            + "If the vertical coordinate is not needed, set surface_only=True"
+            "the pressure. This can be done either with PMID or with"
+            "P0, PS, hyam and hybm."
+            "If the vertical coordinate is not needed, set surface_only=True"
```
```suggestion
            "the pressure. This can be done either with PMID or with "
            "P0, PS, hyam and hybm. "
            "If the vertical coordinate is not needed, set surface_only=True"
```

zmoon commented on monetio/models/_cesm_fv_mm.py:
```diff
@@ -183,9 +195,9 @@ def _calc_pressure(dset):
     if not all(pvar in list(dset.keys()) for pvar in presvars):
         raise KeyError(
             "The model does not have the variables to calculate"
```
```suggestion
            "The model does not have the variables to calculate "
```

zmoon commented:
> maybe we can come up with a new test file/case by selecting a few levels from data [here](https://csl.noaa.gov/groups/csl4/modeldata/melodies-monet/data/example_model_data/cesmfv_example/) since currently we're only testing a surf-only case.

@noribeth-m what do you think about this? Do you have a preference regarding file to use? Do you want to give adding the new test case a go? Or I could do it.

noribeth-m committed:
Update monetio/models/_cesm_fv_mm.py

Co-authored-by: Zachary Moon <zmoon92@gmail.com>

noribeth-m committed:
Update monetio/models/_cesm_fv_mm.py

Co-authored-by: Zachary Moon <zmoon92@gmail.com>

noribeth-m commented:
> > maybe we can come up with a new test file/case by selecting a few levels from data [here](https://csl.noaa.gov/groups/csl4/modeldata/melodies-monet/data/example_model_data/cesmfv_example/) since currently we're only testing a surf-only case.
> 
> @noribeth-m what do you think about this? Do you have a preference regarding file to use? Do you want to give adding the new test case a go? Or I could do it.

I think this is a good idea. I would like to try and create a new test file using data that includes multiple levels. I'll give it a go now and let you know if I'm able to make any progress.

noribeth-m committed:
Fix time dimension handling in the hybrid variables in pressure calculations for multiple model outputs because when more than one model file is used, a time dimension is created.

noribeth-m committed:
Created a test file for _cesm_fv_mm.py (see notes for a successful run in pull request #228)

noribeth-m committed:
Merge branch 'develop_nori' of github.com:bgaubertNCAR/monetio into develop_nori

noribeth-m commented:
Hi @zmoon -- Sorry for the delay. I was able to set up an updated test file for some of the changes I made to _cesm_fv_mm.py, which I've committed to develop_nori. 

A couple issues that I ran into while testing: 

1. I attempted to use this test file -- CAM_chem_merra2_FCSD_1deg_QFED_world_201909-01-09_small.nc -- because it has the vertical levels. Some of the tests that I wrote don't work with this file because there are variables that are missing such as temperature ('T') and surface geopotential ('PHIS'), which are needed to test the calculation of the hydrostatic height, which is then used to calculate the layer thickness ('dz_m'). For testing, I used a model output file that I had, and all of the tests passed. 
2. The _cesm_fv_mm.py code, when giving multiple model paths, is adding a time dimension to the hybrid variables. I added a fix for dealing with this in the reader.

noribeth-m committed:
Added the function _calc_layer_thickness_mid() that uses PMID and PDELDRY. Added warnings to skip reading some of the variables that might not be needed with this new function like PS and PHIS.

noribeth-m committed:
had PS renamed incorrectly, corrected this!

noribeth-m committed:
Using the test file I provide, it does not contain the default variables, so I specify the var_list for the tests in open_mfdataset.

noribeth-m committed:
Corrected a variable name, that it was not reading in. Cleaned up the code as well.

noribeth-m committed:
removed surfpres_pa and alt_agl_m_mid from testing because they are not saved after finishing opening in open_mfdataset.

noribeth-m committed:
updated ftp file name to read in cam-chem run with variables needed for testing.

noribeth-m committed:
corrected the file path name to a string

rschwant commented:
Looks like there is now just an issue with downloading the AERONET data. Likely there was a brief moment where the data was not available. Can we try re-running the test? I don't have authority to do this, so I think @bbakernoaa only you can.

noribeth-m committed:
removed some empty spaces

noribeth-m commented:
Hi @bbakernoaa and @rschwant, I looked through the log file and ran the test on my end as well, and it continues failing due to test_aeronet.py, when it attempts to load data from the URL. I checked the URL and it is accessible, so I am unsure why this is happening.

bbakernoaa commented:
@noribeth-m @rschwant I went ahead and approved this.  I'm working on an AERONET fix to fix the CI issues
